wallet-new/src/Cardano/Wallet/API/Request/Pagination.hs

```diff
@@ -0,0 +1,102 @@
+{- | Support for resource pagination.
+-}
+{-# LANGUAGE DeriveGeneric              #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE RankNTypes                 #-}
+module Cardano.Wallet.API.Request.Pagination (
+      Page(..)
+    , PerPage(..)
+    , maxPerPageEntries
+    , defaultPerPageEntries
+    , PaginationMetadata(..)
+    , PaginationParams(..)
+    ) where
+
+import           Universum
+
+import           Data.Aeson.TH
+import           Data.Default
+import qualified Serokell.Aeson.Options as Serokell
+import           Test.QuickCheck (Arbitrary (..), choose, getPositive)
+import           Web.HttpApiData
+
+-- | A `Page` is used in paginated endpoints to request access to a particular
+-- subset of a collection.
+newtype Page = Page Int
+             deriving (Show, Eq, Ord, Num)
+
+deriveJSON Serokell.defaultOptions ''Page
+
+instance Arbitrary Page where
+  arbitrary = Page . getPositive <$> arbitrary
+
+instance FromHttpApiData Page where
+    parseQueryParam qp = case parseQueryParam qp of
+        Right (p :: Int) | p < 1 -> Left "A page number cannot be less than 1."
+        Right (p :: Int) -> Right (Page p)
+        Left e           -> Left e
+
+instance ToHttpApiData Page where
+    toQueryParam (Page p) = fromString (show p)
+
+-- | If not specified otherwise, return first page.
+instance Default Page where
+    def = Page 1
+
+-- | A `PerPage` is used to specify the number of entries which should be returned
+-- as part of a paginated response.
+newtype PerPage = PerPage Int
```

Comment:
Also maybe `Word64` ?

Reply:
@akegalj Do you really think we might want to send more than `2147483648` entries in a _single_ page? 😉